### ISSUE TITLE
Use a less ancient OpenSSL method in twisted.test.test_sslverify.OpenSSLOptionsTests.test_certificateOptionsSerialization

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8187]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8187|
|Type|defect|
|Created|2016-01-29 01:00:47Z|
|Branch|https://github.com/twisted/twisted/tree/less-ancient-method-8187|

Currently this uses `SSLv3_METHOD` which no longer exists in Debian unstable. There's no particular reason to use such an old method, `TLSv1_2_METHOD` should be just fine.

<details><summary>Searchable metadata</summary>

```
trac-id__8187 8187
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__branches_less_ancient_method_8187 branches/less-ancient-method-8187
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1454029247680976 1454029247680976
changetime__1454300806403948 1454300806403948
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_1"></a>|@mithrandi commented|
|-|-|

(In [46684]) Branching to less-ancient-method-8187.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi removed owner|
|-|-|

Buildbot is green: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/less-ancient-method-8187

I ended up using `SSLv23_METHOD` because that one basically has to be present everywhere (turns out some of the supported platforms are too old to have `TLSv1_2_METHOD`...)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @mithrandi|
|-|-|

Thanks for the update.

It looks good... but is missing the news file fragment.

I would say that we need a new buildslave to cover latest Debian unstable... as otherwise Debian unstable would be not supported.

Can you please follow up over the mailing list to see if we can find someone to host Debian Unstable as a slave?

Looking at https://www.openssl.org/docs/manmaster/ssl/SSL_CTX_new.html I see that SSLv23_method is also deprecated.

Where did you found the info that `SSLv23_METHOD because that one basically has to be present everywhere` ?

Can we write the test to first check if TLS_METHOD is available and fallback to SSLv23_METHOD ?

I assume that with SSLv23_METHOD we will need to upgrade it again soon.

As long as the current supported buildslaves are green ... and your systems are also able to run the test, I am happy to have this patch merged in trunk :)

So only minor comments. Please check my comments and merge at will :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_5"></a>|@mithrandi commented|
|-|-|

* I added a news fragment (just .misc, since this is just a test change).
* Started a thread on the mailing list about a Debian unstable builder.
* I didn't know about (or rather, I think I forgot about) the new TLS_METHOD; however, this is new in OpenSSL 1.1.0 which is not yet released, and neither PyOpenSSL nor Python 3 stdlib ssl even have bindings for this new method, so I don't think we'll have to worry about the old SSLv23_METHOD going away any time soon. Trying TLS_METHOD first is tricky when I don't know for sure what the new binding will even be called, once it exists. As far as eventually switching over goes, there's a lot of code that would need to be changed, and a simple search would find this bit along with the rest of it.
* And to make it explicit, this test does pass for me locally on Debian unstable with this branch.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_6"></a>|@mithrandi set status to `closed`|
|-|-|

(In [46698]) Merge less-ancient-method-8187: Use an OpenSSL method that is more widely present for testing purposes.

Author: mithrandi
Reviewer: adiroban
Fixes: [#8187](https://github.com/twisted/twisted/issues/8187)

twisted.test.test_sslverify.OpenSSLOptionsTests.test_certificateOptionsSerialization
previously used SSLv3_METHOD which Debian (and others) are compiling OpenSSL
without these days; SSLv23_METHOD is now used for greater universality.
